src/v/archival/segment_reupload.cc

```diff
@@ -140,6 +140,16 @@ void segment_collector::do_collect(segment_collector_mode mode) {
         }
 
         auto segment_size = result.segment->size_bytes();
+
+        // This is a hack! If the start offset of the range lies inside
+        // the first segment, we treat this segment as free for the purpose
+        // of size accounting. Otherwise, we could use too much of the size
+        // allocation for offests that are not in the range and fall through the
+        // second if statement below before reaching the end of the range.
+        if (result.segment->offsets().base_offset < start) {
```

Comment:
If I understand the change correctly, this comparison can be more clearly understandable as `result.segment->offsets().base_offset < _begin_inclusive`. 

`start` is a moving offset and after the first segment is collected and `start` is updated, this expression `result.segment->offsets().base_offset < start` should never be true.

Reply:
I would suggest it should be rewritten as `result.segment->offsets().base_offset < _begin_inclusive` if possible.

Reply:
done

Reply:
Interestingly, another test started to fail with this change.